core/src/main/scala/org/apache/spark/SparkEnv.scala

```diff
@@ -71,6 +69,10 @@ class SparkEnv (
     val outputCommitCoordinator: OutputCommitCoordinator,
     val conf: SparkConf) extends Logging {
 
+  // We initialize the ShuffleManager later, in SparkContext and Executor, to allow
+  // user jars to define custom ShuffleManagers.
+  var shuffleManager: ShuffleManager = _
```

Comment:
Given `SparkEnv` is a `DeveloperApi`, let us not expose this for mutation.

```suggestion
  private var _shuffleManager: ShuffleManager = _
  
  def shuffleManager: ShuffleManager = _shuffleManager
```

Reply:
Will fix !!